File: internal/tensor/parallel.go
package tensor

import (
	"runtime"
	"sync"
)

// minParallel is the work size below which an op runs serially — goroutine
// overhead isn't worth it for small tensors (typical training parameters).
const minParallel = 8192

// runChunks splits [0, n) into contiguous chunks across `workers` goroutines and
// waits for them. body(lo, hi) must write only outputs in [lo, hi), so the
// result is race-free and bit-identical to a serial run.
func runChunks(n, workers int, body func(lo, hi int)) {
	if workers < 2 || n < 2 {
		body(0, n)
		return
	}
	if workers > n {
		workers = n
	}
	chunk := (n + workers - 1) / workers
	var wg sync.WaitGroup
	for lo := 0; lo < n; lo += chunk {
		hi := lo + chunk
		if hi > n {
			hi = n
		}
		wg.Add(1)
		go func(lo, hi int) {
			defer wg.Done()
			body(lo, hi)
		}(lo, hi)
	}
	wg.Wait()
}

// parallelFor runs body over [0, n) across cores when n is large enough that
// the parallelism pays off. Parallelism never changes a program's output.
func parallelFor(n int, body func(lo, hi int)) {
	workers := 1
	if n >= minParallel {
		workers = runtime.GOMAXPROCS(0)
	}
	runChunks(n, workers, body)
}

// workers returns the goroutine count to use for a task of the given total work.
func workersFor(work int) int {
	if work < minParallel {
		return 1
	}
	return runtime.GOMAXPROCS(0)
}

// sumChunk is the fixed block size for parallelSum. Blocking by a fixed size
// (not by core count) makes the result deterministic regardless of how many
// goroutines run, and is slightly more accurate than a naive running sum.
const sumChunk = 4096

// parallelSum adds a slice deterministically: fixed-size blocks are summed
// independently, then their partials are combined in order. The result is the
// same on any number of cores.
func parallelSum(data []float64) float64 {
	n := len(data)
	if n < minParallel {
		s := 0.0
		for _, x := range data {
			s += x
		}
		return s
	}
	nblocks := (n + sumChunk - 1) / sumChunk
	partials := make([]float64, nblocks)
	runChunks(nblocks, workersFor(n), func(lo, hi int) {
		for blk := lo; blk < hi; blk++ {
			start := blk * sumChunk
			end := start + sumChunk
			if end > n {
				end = n
			}
			s := 0.0
			for i := start; i < end; i++ {
				s += data[i]
			}
			partials[blk] = s
		}
	})
	total := 0.0
	for _, p := range partials {
		total += p
	}
	return total
}

File: rust/src/parallel.rs
//! Threading, ported from `internal/tensor/parallel.go`.
//!
//! The contract carried over verbatim: chunking never changes a program's
//! output. Every body here writes only its own disjoint output range, and the
//! one reduction that could not honour that (a sum) blocks by a fixed size
//! instead of by core count, so the answer does not depend on the machine.

use std::thread::available_parallelism;

/// Work size below which an op runs serially. Thread setup is not worth it for
/// tensors the size of a typical training parameter.
pub const MIN_PARALLEL: usize = 8192;

/// Fixed block size for [`parallel_sum`]. Fixed rather than core-derived so the
/// result is bit-identical on any machine, which costs a little parallel
/// efficiency on very wide hosts and buys the reproducibility guarantee.
pub const SUM_CHUNK: usize = 4096;

pub fn num_workers() -> usize {
    available_parallelism().map(|n| n.get()).unwrap_or(1)
}

/// Worker count for a task of the given total work.
pub fn workers_for(work: usize) -> usize {
    if work < MIN_PARALLEL {
        1
    } else {
        num_workers()
    }
}

/// Splits `n` logical units across `workers` scoped threads. `out` holds
/// `stride` elements per unit and is split so each thread owns its slice
/// outright; that is what makes the race-freedom a compile-time fact here
/// rather than a comment as it is in the Go original.
pub fn run_chunks<F>(n: usize, workers: usize, stride: usize, out: &mut [f64], body: F)
where
    F: Fn(usize, &mut [f64]) + Sync,
{
    if workers < 2 || n < 2 {
        body(0, out);
        return;
    }
    let w = workers.min(n);
    let chunk = n.div_ceil(w);
    std::thread::scope(|s| {
        let mut lo = 0usize;
        let mut rest = out;
        while lo < n {
            let hi = (lo + chunk).min(n);
            let (head, tail) = rest.split_at_mut((hi - lo) * stride);
            rest = tail;
            let b = &body;
            s.spawn(move || b(lo, head));
            lo = hi;
        }
    });
}

/// Runs `body` over `[0, n)` across cores when `n` is large enough to pay for
/// it. `out` must have exactly `n` elements; `body` receives the chunk's start
/// index and the chunk itself.
pub fn parallel_for<F>(n: usize, out: &mut [f64], body: F)
where
    F: Fn(usize, &mut [f64]) + Sync,
{
    let workers = if n >= MIN_PARALLEL { num_workers() } else { 1 };
    run_chunks(n, workers, 1, out, body);
}

/// Adds a slice deterministically: fixed-size blocks summed independently, then
/// their partials combined in index order.
pub fn parallel_sum(data: &[f64]) -> f64 {
    let n = data.len();
    if n < MIN_PARALLEL {
        let mut s = 0.0;
        for &x in data {
            s += x;
        }
        return s;
    }
    let nblocks = n.div_ceil(SUM_CHUNK);
    let mut partials = vec![0.0; nblocks];
    run_chunks(nblocks, workers_for(n), 1, &mut partials, |lo, slots| {
        for (bi, slot) in slots.iter_mut().enumerate() {
            let start = (lo + bi) * SUM_CHUNK;
            let end = (start + SUM_CHUNK).min(n);
            let mut s = 0.0;
            for &x in &data[start..end] {
                s += x;
            }
            *slot = s;
        }
    });
    let mut total = 0.0;
    for &p in &partials {
        total += p;
    }
    total
}
